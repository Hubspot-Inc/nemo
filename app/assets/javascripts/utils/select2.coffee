select2 = {
  getAjaxParams: (url, resultsKey = 'results', textKey = 'text') ->
    url: url
    dataType: 'json'
    delay: 250,
    data: (params) ->
      search: params.term
      page: params.page
    processResults: (data) ->
      results = data[resultsKey]
      results.forEach((r) -> r['text'] = r[textKey]) unless textKey == 'text'
      {results: results, pagination: {more: data.more}}
    cache: true
}

ELMO.select2 = select2
