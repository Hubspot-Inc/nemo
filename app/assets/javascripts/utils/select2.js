export function select2AjaxParams(url, resultsKey = 'results') {
  return {
    url,
    dataType: 'json',
    delay: 250,
    data: ({ term: search, page }) => ({
      search,
      page,
    }),
    processResults: ({ [resultsKey]: results, more }) => ({
      results,
      pagination: { more },
    }),
    cache: true,
  };
}
