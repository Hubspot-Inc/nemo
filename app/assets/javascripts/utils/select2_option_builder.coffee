# Builds options for select2 controls so that common options can be reused without duplication.
# It's designed like this, instead of as a wrapper of select2, so it can be used in React as well
# as Backbone.
class ELMO.Utils.Select2OptionBuilder
  ajax: (url, resultsKey = 'results', textKey = 'text') ->
    url: url
    dataType: 'json'
    delay: 250,
    data: (params) ->
      search: params.term
      page: params.page
    processResults: (data) ->
      results = data[resultsKey]
      results.forEach((r) -> r['text'] = r[textKey]) unless textKey == 'text'
      {results: results, pagination: {more: data.more}}
    cache: true
