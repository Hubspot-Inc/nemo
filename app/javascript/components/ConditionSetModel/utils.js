import ConditionSetModel from './ConditionSetModel';

/** Cache. */
const conditionSetStores = {};

/**
 * Returns a new instance of ConditionSetModel.
 *
 * Generally this should be added to a top-level Provider and used
 * once per condition set.
 */
export function provideConditionSetStore(uniqueId) {
  if (!conditionSetStores[uniqueId]) {
    conditionSetStores[uniqueId] = new ConditionSetModel();

    if (process.env.NODE_ENV === 'development') {
      // Debug helper.
      window.store = window.store || {};
      window.store[uniqueId] = conditionSetStores[uniqueId];
    }
  }

  return conditionSetStores[uniqueId];
}
