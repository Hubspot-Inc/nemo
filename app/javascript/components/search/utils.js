import isEmpty from 'lodash/isEmpty';
import queryString from 'query-string';

import FiltersModel from './FiltersModel';
import QuestionFilterModel from './QuestionFilterModel';

const MAX_HINTS_BEFORE_ELLIPSIZE = 1;

/**
 * Stringified controller_name from Rails.
 */
export const CONTROLLER_NAME = {
  RESPONSES: '"responses"',
};

/**
 * Returns a new instance of FiltersModel.
 *
 * Generally this should be added to a top-level Provider and only created once.
 */
export function createFiltersStore() {
  const store = new FiltersModel();

  if (process.env.NODE_ENV === 'development') {
    // Debug helper.
    if (window.store) {
      console.warn('WARN: Trying to create store that already exists.');
    }
    window.store = store;
  }

  return store;
}

/**
 * Returns a new instance of QuestionFilterModel.
 *
 * Generally this should be added to a top-level Provider and created
 * once per group of questions.
 */
export function createQuestionFilterStore(debugName) {
  const store = new QuestionFilterModel();

  if (process.env.NODE_ENV === 'development') {
    // Debug helper.
    window.store = window.store || {};
    if (window.store[debugName]) {
      console.warn('WARN: Trying to create store that already exists:', debugName);
    }
    window.store[debugName] = store;
  }

  return store;
}

/**
 * Given a list of hints (e.g. currently selected form names for the form filter button),
 * stringify them to be displayed on the button itself.
 */
export function getButtonHintString(hints) {
  if (isEmpty(hints)) {
    return '';
  }

  const joinedHints = hints.length > MAX_HINTS_BEFORE_ELLIPSIZE
    ? hints.length
    : hints.join(', ');

  return ` (${joinedHints})`;
}

/**
 * Given a form ID, find it in the list of all forms and return the name.
 */
export function getFormNameFromId(allForms, searchId) {
  const form = allForms.find(({ id }) => searchId === id);
  return (form && form.name) || 'Unknown';
}

/**
 * Given all of the different filter states,
 * return a stringified version for the backend.
 */
export function getFilterString({ allForms, selectedFormIds, advancedSearchText }) {
  const selectedFormNames = selectedFormIds
    .map((id) => JSON.stringify(getFormNameFromId(allForms, id)));

  const parts = [
    isEmpty(selectedFormNames) ? null : `form:(${selectedFormNames.join('|')})`,
    advancedSearchText,
  ].filter(Boolean);

  return parts.join(' ');
}

/**
 * Reload the page with the given search.
 */
export function submitSearch(filterString) {
  const parsed = queryString.parse(window.location.search);
  // The `search` query param will be removed from the URL if it's `undefined`.
  const search = filterString || undefined;
  const params = queryString.stringify({ ...parsed, search });

  window.location.assign(params
    ? `?${params}`
    : window.location.pathname);
}

/**
 * Returns true if the given param name exists and is non-empty.
 */
export function isQueryParamTruthy(paramName) {
  const parsed = queryString.parse(window.location.search);
  return Boolean(parsed[paramName]);
}
