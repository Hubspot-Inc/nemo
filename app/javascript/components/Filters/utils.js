import isEmpty from 'lodash/isEmpty';
import mapKeys from 'lodash/mapKeys';
import queryString from 'query-string';
import moment from 'moment';
import { last } from 'lodash';

import { SUBMITTER_TYPES } from './SubmitterFilter/utils';

const MAX_HINTS_BEFORE_ELLIPSIZE = 1;

/**
 * Stringified controller_name from Rails.
 */
export const CONTROLLER_NAME = {
  RESPONSES: '"responses"',
};

/** Cache. */
let filtersStore = null;

/**
 * Returns a new instance of FiltersModel.
 *
 * Generally this should be added to a top-level Provider and only used once.
 */
export function provideFiltersStore(FiltersModel, initialState) {
  if (!filtersStore) {
    filtersStore = new FiltersModel(initialState);

    if (process.env.NODE_ENV === 'development') {
      // Debug helper.
      window.store = filtersStore;
    }
  }

  return filtersStore;
}

/**
 * Given a list of hints (e.g. currently selected form names for the form filter button),
 * stringify them to be displayed on the button itself.
 */
export function getButtonHintString(hints) {
  if (isEmpty(hints)) {
    return '';
  }

  const joinedHints = hints.length > MAX_HINTS_BEFORE_ELLIPSIZE
    ? `${hints.length} filters`
    : hints.join(', ');

  return ` (${joinedHints})`;
}

/**
 * Given an item ID, find it in the list of items and return its `name`.
 */
export function getItemNameFromId(allItems, searchId, nameKey = 'name') {
  const item = allItems.find(({ id }) => searchId === id);
  return (item && item[nameKey]) || 'Unknown';
}

/**
 * Given a leftQingId, find it in the list of all questions and return the name.
 */
export function getQuestionNameFromId(allQuestions, searchId) {
  return getItemNameFromId(allQuestions, searchId, 'code');
}

/**
 * Converts a list of data from the backend into something Select2 understands, e.g.
 * [{ id: '1', name: 'One' }, ...] => [{ id: '1', text: 'One' }, ...]
 */
export function parseListForSelect2(allItems) {
  return allItems.map((item) =>
    mapKeys(item, (value, key) => (key === 'name' ? 'text' : key)));
}

/**
 * Given all of the different filter states,
 * return a stringified version for the backend.
 */
export function getFilterString({
  selectedFormIds,
  conditionSetStore,
  isReviewed,
  selectedSubmittersForType,
  advancedSearchText,
  startDate,
  endDate,
}) {
  const allQuestions = conditionSetStore.refableQings;
  const questionFilters = conditionSetStore.conditions
    .filter(({ leftQingId, currTextValue, remove }) => leftQingId && currTextValue && !remove)
    .map(({ leftQingId, currTextValue }) =>
      `{${getQuestionNameFromId(allQuestions, leftQingId)}}:${JSON.stringify(currTextValue)}`);

  const submitterParts = SUBMITTER_TYPES.map((type) => {
    const selectedSubmitterIds = selectedSubmittersForType[type].map(({ id }) => id);
    return isEmpty(selectedSubmitterIds) ? null : `${type}-id:(${selectedSubmitterIds.join('|')})`;
  });

  // Only include one upper bound and one lower bound date. Use widest range.
  let searchDates = advancedSearchText.split(' ').filter((s) => s.indexOf('submit-date') !== -1).map(queryToMoment);
  const searchText = advancedSearchText.split(' ').filter((s) => s.indexOf('submit-date') === -1).join(' ');
  searchDates = [...searchDates, startDate, endDate].filter((d) => d !== null);
  if (searchDates.length >= 2) {
    searchDates.sort((a, b) => {
      if (a.isSameOrBefore(b)) return -1;
      return 1;
    });
  }

  let dateQueries = null;
  if (searchDates.length === 0) {
    // No-op
  } else if (searchDates.length === 1) {
    // Figure out if this was supposed to be an upper or lower bound.
    if (searchDates[0] === startDate) {
      dateQueries = `submit-date >= ${searchDates[0].format('YYYY-MM-DD')}`;
    } else if (searchDates[0] === endDate) {
      dateQueries = `submit-date <= ${searchDates[0].format('YYYY-MM-DD')}`;
    } else {
      // Date entered in search bar, use original string.
      dateQueries = advancedSearchText.split(' ').filter((s) => s.indexOf('submit-date') !== -1).join(' ');
    }
  } else {
    // Take first and last dates.
    dateQueries = `submit-date >= ${searchDates[0].format('YYYY-MM-DD')} submit-date <= ${last(searchDates).format('YYYY-MM-DD')}`;
  }

  const parts = [
    isEmpty(selectedFormIds) ? null : `form-id:(${selectedFormIds.join('|')})`,
    ...questionFilters,
    isReviewed == null ? null : `reviewed:${isReviewed ? '1' : '0'}`,
    ...submitterParts,
    searchText,
    dateQueries,
  ].filter(Boolean);

  return parts.join(' ');
}

/**
 * Reload the page with the given search.
 */
export function submitSearch(filterString) {
  const parsed = queryString.parse(window.location.search);
  // The `search` query param will be removed from the URL if it's `undefined`.
  const search = filterString || undefined;
  const params = queryString.stringify({ ...parsed, search });

  window.location.assign(params
    ? `?${params}`
    : window.location.pathname);
}

/**
 * Returns true if the given param name exists and is non-empty.
 */
export function isQueryParamTruthy(paramName) {
  const parsed = queryString.parse(window.location.search);
  return Boolean(parsed[paramName]);
}

export function queryToMoment(query) {
  if (query === null || query === undefined) return null;
  const dateString = query.split(/submit-date[<>=]+/)[1];
  // If this doesn't parse an actual date, don't use it.
  if (dateString === undefined) return null;
  const date = moment(dateString);
  if (date._isValid) { /* eslint-disable-line no-underscore-dangle */
    return date;
  }
  return null;
}
