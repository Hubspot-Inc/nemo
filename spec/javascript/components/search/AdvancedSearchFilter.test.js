import React from "react";
import {shallow} from "enzyme";

import {advancedSearchProps} from "./utils";

import Component from "../../../../app/javascript/components/search/AdvancedSearchFilter";

const defaultProps = {
  ...advancedSearchProps,
  onChangeAdvancedSearch: jest.fn(),
  onSubmit: jest.fn(),
  onClear: jest.fn(),
};

function withQueryParam() {
  window.location.search = "?search=foo";
}

function withoutQueryParam() {
  window.location.search = "?search=";
}

it("renders as expected (without query param)", () => {
  withoutQueryParam();
  const wrapper = shallow(<Component {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders as expected (with query param)", () => {
  withQueryParam();
  const wrapper = shallow(<Component {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

describe("callbacks", () => {
  withQueryParam();
  const wrapper = shallow(<Component {...defaultProps} />);

  it("submits", () => {
    wrapper.find("Button.btn-apply").simulate("click");
    expect(defaultProps.onSubmit).toMatchSnapshot();
  });

  it("clears", () => {
    wrapper.find("Button.btn-clear").simulate("click");
    expect(defaultProps.onClear).toMatchSnapshot();
  });
});
