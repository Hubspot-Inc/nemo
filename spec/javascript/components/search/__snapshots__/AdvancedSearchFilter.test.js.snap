// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callbacks clears 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`callbacks submits 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`renders as expected (with query param) 1`] = `
<div>
  <input
    autoComplete="off"
    className="form-control"
    id="search_str"
    name="search"
    onChange={[MockFunction]}
    placeholder="Advanced search"
    type="text"
    value="query"
  />
  <Button
    className="btn-apply btn-advanced-search"
    onClick={[MockFunction]}
  >
    Search
  </Button>
  <Button
    className="btn-clear btn-margin-left"
    onClick={[MockFunction]}
  >
    Clear
  </Button>
</div>
`;

exports[`renders as expected (without query param) 1`] = `
<div>
  <input
    autoComplete="off"
    className="form-control"
    id="search_str"
    name="search"
    onChange={[MockFunction]}
    placeholder="Advanced search"
    type="text"
    value="query"
  />
  <Button
    className="btn-apply btn-advanced-search"
    onClick={[MockFunction]}
  >
    Search
  </Button>
</div>
`;
